File: lib/pull_preview/instance.rb
require "erb"
require "ostruct"

module PullPreview
  class Instance
    include Utils

    attr_reader :admins
    attr_reader :cidrs
    attr_reader :compose_files
    attr_reader :compose_options
    attr_reader :default_port
    attr_reader :dns
    attr_reader :name
    attr_reader :subdomain
    attr_reader :ports
    attr_reader :provider
    attr_reader :registries
    attr_reader :size
    attr_reader :tags
    attr_reader :access_details

    class << self
      attr_accessor :client
      attr_accessor :logger
    end

    def self.normalize_name(name)
      name.
        gsub(/[^a-z0-9]/i, "-").
        squeeze("-")[0..60].
        gsub(/(^-|-$)/, "")
    end

    def initialize(name, opts = {})
      @provider = PullPreview.provider
      @name = self.class.normalize_name(name)
      @subdomain = opts[:subdomain] || name
      @admins = opts[:admins] || []
      @cidrs = opts[:cidrs] || ["0.0.0.0/0"]
      @default_port = opts[:default_port] || "80"
      # TODO: normalize
      @ports = (opts[:ports] || []).push(default_port).push("22").uniq.compact
      @compose_files = opts[:compose_files] || ["docker-compose.yml"]
      @compose_options = opts[:compose_options] || ["--build"]
      @registries = opts[:registries] || []
      @dns = opts[:dns]
      @size = opts[:instance_type]
      @ssh_results = []
      @tags = opts[:tags] || {}
    end

    def launch_and_wait_until_ready!
      @access_details = provider.launch!(name, size: size, user_data: user_data, ports: ports, cidrs: cidrs, tags: tags)
      logger.debug "access_details=#{@access_details.inspect}"
      logger.info "Instance is running public_ip=#{public_ip} public_dns=#{public_dns}"
      wait_until_ssh_ready!
    end

    def create_domain_entry
      logger.info "Adding a domain entry public_dns=#{public_dns} and target #{public_ip} in DNS Zone"
      provider.create_domain_entry(dns, public_dns, public_ip)
    end

    def delete_domain_entry
      logger.info "Deleting a domain entry public_dns=#{public_dns} and target #{public_ip} in DNS Zone"
      provider.delete_domain_entry(dns, public_dns, public_ip)
    end

    def terminate!
      if provider.terminate!(name)
        logger.info "Instance successfully destroyed"
      else
        logger.error "Unable to destroy instance"
      end
    end

    def running?
      provider.running?(name)
    end

    def ssh_ready?
      ssh("test -f /etc/pullpreview/ready")
    end

    def public_ip
      access_details.ip_address
    end

    def public_dns
      reserved_space_for_user_subdomain = 8
      # https://community.letsencrypt.org/t/a-certificate-for-a-63-character-domain/78870/4
      remaining_chars_for_subdomain = 62 - reserved_space_for_user_subdomain - dns.size - public_ip.size - "ip".size - ("." * 3).size
      [
        [subdomain[0..remaining_chars_for_subdomain], "ip", public_ip.gsub(".", "-")].join("-").squeeze("-"),
        dns
      ].join(".")
    end

    def url
      scheme = (default_port == "443" ? "https" : "http")
      "#{scheme}://#{public_dns}"
    end

    def username
      provider.username
    end

    def ssh_public_keys
      @ssh_public_keys ||= admins.map do |github_username|
        URI.open("https://github.com/#{github_username}.keys").read.split("\n")
      end.flatten.reject{|key| key.empty?}
    end

    def user_data
      @user_data ||= UserData.new(app_path: remote_app_path, username: username, ssh_public_keys: ssh_public_keys)
    end

    def erb_locals
      OpenStruct.new(
        remote_app_path: remote_app_path,
        compose_files: compose_files,
        compose_options: compose_options,
        public_ip: public_ip,
        public_dns: public_dns,
        admins: admins,
        url: url,
      )
    end

    def github_token
      ENV.fetch("GITHUB_TOKEN", "")
    end

    def github_repository_owner
      ENV.fetch("GITHUB_REPOSITORY_OWNER", "")
    end

    def update_script
      PullPreview.data_dir.join("update_script.sh.erb")
    end

    def update_script_rendered
      ERB.new(File.read(update_script)).result_with_hash(locals: erb_locals)
    end

    def setup_ssh_access
      File.open("/tmp/authorized_keys", "w+") do |f|
        f.write ssh_public_keys.join("\n")
      end
      scp("/tmp/authorized_keys", "/home/#{username}/.ssh/authorized_keys", mode: "0600")
      # in case provider ssh user is different than the one we want to use
      ssh("chown #{username}.#{username} /home/#{username}/.ssh/authorized_keys && chmod 0600 /home/#{username}/.ssh/authorized_keys")
    end

    def setup_update_script
      tmpfile = Tempfile.new("update_script").tap do |f|
        f.write update_script_rendered
      end
      tmpfile.flush
      unless scp(tmpfile.path, "/tmp/update_script.sh", mode: "0755")
        raise Error, "Unable to copy the update script on instance. Aborting."
      end
    end

    def setup_prepost_scripts
      tmpfile = Tempfile.new(["prescript", ".sh"])
      tmpfile.puts "#!/bin/bash -e"
      registries.each_with_index do |registry, index|
        begin
          uri = URI.parse(registry)
          raise Error, "Invalid registry" if uri.host.nil? || uri.scheme != "docker"
          username = uri.user
          password = uri.password
          if password.nil?
            password = username
            username = "doesnotmatter"
          end
          tmpfile.puts 'echo "Logging into %{host}..."' % { host: uri.host }
          # https://docs.github.com/en/packages/guides/using-github-packages-with-github-actions#upgrading-a-workflow-that-accesses-ghcrio
          tmpfile.puts 'echo "%{password}" | docker login "%{host}" -u "%{username}" --password-stdin' % {
            host: uri.host,
            username: username,
            password: password,
          }
        rescue URI::Error, Error => e
          logger.warn "Registry ##{index} is invalid: #{e.message}"
        end
      end
      tmpfile.flush
      unless scp(tmpfile.path, "/tmp/pre_script.sh", mode: "0755")
        raise Error, "Unable to copy the pre script on instance. Aborting."
      end
    end

    def wait_until_ssh_ready!
      if wait_until { logger.info "Waiting for ssh" ; ssh_ready? }
        logger.info "Instance ssh access OK"
      else
        logger.error "Instance ssh access KO"
        raise Error, "Can't connect to instance over SSH. Aborting."
      end
    end

    def scp(source, target, mode: "0644")
      ssh("cat - > #{target} && chmod #{mode} #{target}", input: File.new(source))
    end

    def ssh(command, input: nil)
      key_file_path = "/tmp/tempkey"
      cert_key_path = "/tmp/tempkey-cert.pub"
      File.open(key_file_path, "w+") do |f|
        f.puts access_details.private_key
      end
      if access_details.cert_key
        File.open(cert_key_path, "w+") do |f|
          f.puts access_details.cert_key
        end
      end
      [key_file_path].each{|file| FileUtils.chmod 0600, file}

      cmd = "ssh #{"-v " if logger.level == Logger::DEBUG}-o ServerAliveInterval=15 -o IdentitiesOnly=yes -i #{key_file_path} #{ssh_address} #{ssh_options.join(" ")} '#{command}'"
      if input && input.respond_to?(:path)
        cmd = "cat #{input.path} | #{cmd}"
      end
      logger.debug cmd
      system(cmd).tap {|result| @ssh_results.push([cmd, result])}
    end

    def ssh_address
      access_details.ssh_address
    end

    def ssh_options
      [
        "-o StrictHostKeyChecking=no",
        "-o UserKnownHostsFile=/dev/null",
        "-o LogLevel=ERROR",
        "-o ConnectTimeout=10",
      ]
    end

    private def logger
      PullPreview.logger
    end

    private def remote_app_path
      REMOTE_APP_PATH
    end
  end
end

File: internal/pullpreview/instance.go
package pullpreview

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"time"
)

const (
	remoteAppPath              = "/app"
	instanceSSHReadyInterval   = 5 * time.Second
	instanceSSHReadyWaitWindow = 5 * time.Minute
	sshReadyDiagnosticCommand  = `if test -f /etc/pullpreview/ready; then
  echo ready-marker-present
  exit 0
fi
echo ready-marker-missing
if command -v cloud-init >/dev/null 2>&1; then
  echo "-- cloud-init status --"
  sudo -n cloud-init status --long 2>/dev/null || cloud-init status --long 2>/dev/null || sudo -n cloud-init status 2>/dev/null || cloud-init status 2>/dev/null || true
fi
if sudo -n test -f /var/log/cloud-init-output.log 2>/dev/null || test -f /var/log/cloud-init-output.log; then
  echo "-- cloud-init-output tail --"
  sudo -n tail -n 40 /var/log/cloud-init-output.log 2>/dev/null || tail -n 40 /var/log/cloud-init-output.log 2>/dev/null || true
fi
exit 1`
)

type Runner interface {
	Run(cmd *exec.Cmd) error
}

type SystemRunner struct{}

func (r SystemRunner) Run(cmd *exec.Cmd) error {
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	return cmd.Run()
}

var runSSHCombinedOutput = func(cmd *exec.Cmd) ([]byte, error) {
	return cmd.CombinedOutput()
}

type Instance struct {
	Name             string
	Subdomain        string
	Admins           []string
	AdminPublicKeys  []string
	Context          context.Context
	DeploymentTarget DeploymentTarget
	CIDRs            []string
	ComposeFiles     []string
	ComposeOptions   []string
	Chart            string
	ChartRepository  string
	ChartValues      []string
	ChartSet         []string
	DefaultPort      string
	DNS              string
	Ports            []string
	ProxyTLS         string
	Provider         Provider
	Registries       []string
	Size             string
	Tags             map[string]string
	PreScript        string
	Access           AccessDetails
	Logger           *Logger
	Runner           Runner
}

func NewInstance(name string, opts CommonOptions, provider Provider, logger *Logger) *Instance {
	normalized := NormalizeName(name)
	target := NormalizeDeploymentTarget(string(opts.DeploymentTarget))
	defaultPort := defaultString(opts.DefaultPort, "80")
	proxyTLS := strings.TrimSpace(opts.ProxyTLS)
	if proxyTLS != "" {
		if defaultPort != "443" && logger != nil {
			logger.Warnf("proxy_tls=%q enabled: overriding default_port=%s to 443", proxyTLS, defaultPort)
		}
		defaultPort = "443"
	}
	return &Instance{
		Name:             normalized,
		Subdomain:        NormalizeName(name),
		Admins:           opts.Admins,
		AdminPublicKeys:  opts.AdminPublicKeys,
		Context:          ensureContext(opts.Context),
		DeploymentTarget: target,
		CIDRs:            defaultSlice(opts.CIDRs, []string{"0.0.0.0/0"}),
		ComposeFiles:     defaultSlice(opts.ComposeFiles, []string{"docker-compose.yml"}),
		ComposeOptions:   defaultSlice(opts.ComposeOptions, []string{"--build"}),
		Chart:            strings.TrimSpace(opts.Chart),
		ChartRepository:  strings.TrimSpace(opts.ChartRepository),
		ChartValues:      opts.ChartValues,
		ChartSet:         opts.ChartSet,
		DefaultPort:      defaultPort,
		DNS:              defaultString(opts.DNS, "my.preview.run"),
		Ports:            opts.Ports,
		ProxyTLS:         proxyTLS,
		Provider:         provider,
		Registries:       opts.Registries,
		Size:             opts.InstanceType,
		Tags:             defaultMap(opts.Tags),
		PreScript:        opts.PreScript,
		Logger:           logger,
		Runner:           SystemRunner{},
	}
}

func (i *Instance) WithSubdomain(subdomain string) {
	if subdomain == "" {
		return
	}
	i.Subdomain = NormalizeName(subdomain)
}

func defaultSlice(value, fallback []string) []string {
	if len(value) == 0 {
		return fallback
	}
	return value
}

func defaultString(value, fallback string) string {
	if strings.TrimSpace(value) == "" {
		return fallback
	}
	return value
}

func defaultMap(value map[string]string) map[string]string {
	if value == nil {
		return map[string]string{}
	}
	return value
}

func providerName(provider Provider) string {
	if metadata, ok := provider.(ProviderMetadata); ok {
		return strings.ToLower(strings.TrimSpace(metadata.Name()))
	}
	return ""
}

func providerSupportsDeploymentTarget(provider Provider, target DeploymentTarget) bool {
	if supported, ok := provider.(SupportsDeploymentTarget); ok {
		return supported.SupportsDeploymentTarget(target)
	}
	return NormalizeDeploymentTarget(string(target)) == DeploymentTargetCompose
}

func sameStringSlice(got, want []string) bool {
	if len(got) != len(want) {
		return false
	}
	for idx := range got {
		if strings.TrimSpace(got[idx]) != strings.TrimSpace(want[idx]) {
			return false
		}
	}
	return true
}

func (i *Instance) ValidateDeploymentConfig() error {
	if err := i.DeploymentTarget.Validate(); err != nil {
		return err
	}

	switch i.DeploymentTarget {
	case DeploymentTargetCompose:
		if strings.TrimSpace(i.Chart) != "" || strings.TrimSpace(i.ChartRepository) != "" || len(i.ChartValues) > 0 || len(i.ChartSet) > 0 {
			return fmt.Errorf("chart, chart_repository, chart_values, and chart_set require deployment_target=helm")
		}
	case DeploymentTargetHelm:
		if !providerSupportsDeploymentTarget(i.Provider, DeploymentTargetHelm) {
			return fmt.Errorf("deployment_target=helm is unsupported for provider=%s", providerName(i.Provider))
		}
		if strings.TrimSpace(i.Chart) == "" {
			return fmt.Errorf("deployment_target=helm requires chart")
		}
		if strings.TrimSpace(i.ProxyTLS) == "" {
			return fmt.Errorf("deployment_target=helm requires proxy_tls")
		}
		if len(uniqueStrings(i.Registries)) > 0 {
			return fmt.Errorf("registries is unsupported with deployment_target=helm")
		}
		if len(i.ComposeFiles) > 0 && !sameStringSlice(i.ComposeFiles, []string{"docker-compose.yml"}) {
			return fmt.Errorf("compose_files is unsupported with deployment_target=helm")
		}
		if len(i.ComposeOptions) > 0 && !sameStringSlice(i.ComposeOptions, []string{"--build"}) {
			return fmt.Errorf("compose_options is unsupported with deployment_target=helm")
		}
	default:
		return fmt.Errorf("unsupported deployment target %q", i.DeploymentTarget)
	}
	return nil
}

func (i *Instance) LaunchAndWait() error {
	if i.Logger != nil {
		i.Logger.Infof("Creating or restoring instance name=%s size=%s", i.Name, i.Size)
	}

	userData := UserData{
		AppPath:       remoteAppPath,
		SSHPublicKeys: i.SSHPublicKeys(),
		Username:      i.Username(),
	}.Script()
	if provider, ok := i.Provider.(UserDataProvider); ok {
		generatedUserData, err := provider.BuildUserData(UserDataOptions{
			AppPath:          remoteAppPath,
			DeploymentTarget: i.DeploymentTarget,
			SSHPublicKeys:    i.SSHPublicKeys(),
			Username:         i.Username(),
		})
		if err != nil {
			return err
		}
		userData = generatedUserData
	}
	access, err := i.Provider.Launch(i.Name, LaunchOptions{
		Size:     i.Size,
		UserData: userData,
		Ports:    i.PortsWithDefaults(),
		CIDRs:    i.CIDRs,
		Tags:     i.Tags,
	})
	if err != nil {
		return err
	}
	i.Access = access
	if i.Logger != nil {
		i.Logger.Infof(
			"Instance created name=%s public_ip=%s username=%s",
			i.Name,
			i.PublicIP(),
			i.Username(),
		)
	}
	if ok := WaitUntilContext(i.Context, pollAttemptsForWindow(instanceSSHReadyWaitWindow, instanceSSHReadyInterval), instanceSSHReadyInterval, func() bool {
		if i.Logger != nil {
			i.Logger.Infof(
				"Waiting for SSH username=%s ip=%s ssh=\"ssh %s\"",
				i.Username(),
				i.PublicIP(),
				i.SSHAddress(),
			)
		}
		return i.SSHReady()
	}); !ok {
		if i.Logger != nil {
			if diagErr := i.SSHReadyDiagnostic(); diagErr != nil {
				i.Logger.Warnf("SSH readiness diagnostics: %v", diagErr)
			}
		}
		return errors.New("can't connect to instance over SSH")
	}
	if i.Logger != nil {
		i.Logger.Infof("Instance ssh access OK")
	}
	return nil
}

func (i *Instance) Terminate() error {
	return i.Provider.Terminate(i.Name)
}

func (i *Instance) Running() (bool, error) {
	return i.Provider.Running(i.Name)
}

func (i *Instance) SSHReady() bool {
	return i.SSH("test -f /etc/pullpreview/ready", nil) == nil
}

func (i *Instance) SSHReadyDiagnostic() error {
	output, err := i.SSHOutput(sshReadyDiagnosticCommand, nil)
	if err == nil {
		return nil
	}
	output = strings.TrimSpace(output)
	if output == "" {
		return err
	}
	return fmt.Errorf("%w: %s", err, output)
}

func (i *Instance) PublicIP() string {
	return i.Access.IPAddress
}

func (i *Instance) PublicDNS() string {
	return PublicDNS(i.Subdomain, i.DNS, i.PublicIP())
}

func (i *Instance) URL() string {
	scheme := "http"
	if i.DefaultPort == "443" {
		scheme = "https"
	}
	return fmt.Sprintf("%s://%s:%s", scheme, i.PublicDNS(), i.DefaultPort)
}

func (i *Instance) Username() string {
	if i.Access.Username != "" {
		return i.Access.Username
	}
	return i.Provider.Username()
}

func (i *Instance) PortsWithDefaults() []string {
	proxyTLSEnabled := strings.TrimSpace(i.ProxyTLS) != ""
	ports := []string{}
	for _, port := range i.Ports {
		if proxyTLSEnabled && i.DeploymentTarget != DeploymentTargetHelm && firewallRuleTargetsPort(port, 80) {
			continue
		}
		ports = append(ports, port)
	}
	ports = append(ports, i.DefaultPort, "22")
	return uniqueStrings(ports)
}

func firewallRuleTargetsPort(rule string, port int) bool {
	value := strings.TrimSpace(rule)
	if value == "" {
		return false
	}
	if idx := strings.Index(value, "/"); idx >= 0 {
		value = value[:idx]
	}
	if strings.Contains(value, ":") {
		parts := strings.Split(value, ":")
		if len(parts) > 0 {
			value = parts[len(parts)-1]
		}
	}
	parsed, err := strconv.Atoi(strings.TrimSpace(value))
	if err != nil {
		return false
	}
	return parsed == port
}

func uniqueStrings(values []string) []string {
	seen := map[string]struct{}{}
	result := []string{}
	for _, v := range values {
		v = strings.TrimSpace(v)
		if v == "" {
			continue
		}
		if _, ok := seen[v]; ok {
			continue
		}
		seen[v] = struct{}{}
		result = append(result, v)
	}
	return result
}

func (i *Instance) SSHPublicKeys() []string {
	if len(i.AdminPublicKeys) > 0 {
		return uniqueStrings(i.AdminPublicKeys)
	}

	keys := []string{}
	client := http.Client{Timeout: 10 * time.Second}
	for _, admin := range i.Admins {
		admin = strings.TrimSpace(admin)
		if admin == "" {
			continue
		}
		url := fmt.Sprintf("https://github.com/%s.keys", admin)
		resp, err := client.Get(url)
		if err != nil {
			if i.Logger != nil {
				i.Logger.Warnf("Unable to fetch SSH keys for %s: %v", admin, err)
			}
			continue
		}
		body, err := io.ReadAll(resp.Body)
		_ = resp.Body.Close()
		if err != nil {
			continue
		}
		for _, line := range strings.Split(string(body), "\n") {
			line = strings.TrimSpace(line)
			if line != "" {
				keys = append(keys, line)
			}
		}
	}
	return keys
}

func (i *Instance) SetupSSHAccess() error {
	keys := i.SSHPublicKeys()
	if len(keys) == 0 {
		return nil
	}
	content := strings.Join(keys, "\n") + "\n"
	homeDir := HomeDirForUser(i.Username())
	return i.appendRemoteFile(bytes.NewBufferString(content), fmt.Sprintf("%s/.ssh/authorized_keys", homeDir), "0600")
}

func (i *Instance) SetupPreScript() error {
	script := BuildPreScript(i.Registries, i.PreScript, i.Logger)
	return i.SCP(bytes.NewBufferString(script), "/tmp/pre_script.sh", "0755")
}

func (i *Instance) SCP(input io.Reader, target, mode string) error {
	command := fmt.Sprintf("cat - > %s && chmod %s %s", target, mode, target)
	return i.SSH(command, input)
}

func (i *Instance) appendRemoteFile(input io.Reader, target, mode string) error {
	command := fmt.Sprintf("cat - >> %s && chmod %s %s", target, mode, target)
	return i.SSH(command, input)
}

func (i *Instance) sshArgs(keyFile, certFile string) []string {
	args := []string{}
	if i.Logger != nil && i.Logger.level <= LevelDebug {
		args = append(args, "-v")
	}
	args = append(args,
		"-o", "ServerAliveInterval=15",
		"-o", "IdentitiesOnly=yes",
		"-i", keyFile,
	)
	if strings.TrimSpace(certFile) != "" {
		args = append(args, "-o", "CertificateFile="+certFile)
	}
	args = append(args, i.SSHOptions()...)
	args = append(args, i.SSHAddress())
	return args
}

func (i *Instance) SSH(command string, input io.Reader) error {
	keyFile, certFile, err := i.writeTempKeys()
	if err != nil {
		return err
	}
	defer func() {
		_ = os.Remove(keyFile)
		if certFile != "" {
			_ = os.Remove(certFile)
		}
	}()

	args := i.sshArgs(keyFile, certFile)
	args = append(args, command)

	cmd := exec.CommandContext(i.Context, "ssh", args...)
	cmd.Stdin = input
	return i.Runner.Run(cmd)
}

func (i *Instance) SSHOutput(command string, input io.Reader) (string, error) {
	keyFile, certFile, err := i.writeTempKeys()
	if err != nil {
		return "", err
	}
	defer func() {
		_ = os.Remove(keyFile)
		if certFile != "" {
			_ = os.Remove(certFile)
		}
	}()

	args := i.sshArgs(keyFile, certFile)
	args = append(args, command)

	cmd := exec.CommandContext(i.Context, "ssh", args...)
	cmd.Stdin = input
	output, err := runSSHCombinedOutput(cmd)
	return string(output), err
}

func (i *Instance) SSHAddress() string {
	username := i.Username()
	if username == "" {
		return i.PublicIP()
	}
	return fmt.Sprintf("%s@%s", username, i.PublicIP())
}

func (i *Instance) SSHOptions() []string {
	return []string{
		"-o", "BatchMode=yes",
		"-o", "IdentityAgent=none",
		"-o", "StrictHostKeyChecking=no",
		"-o", "UserKnownHostsFile=/dev/null",
		"-o", "LogLevel=ERROR",
		"-o", "ConnectTimeout=10",
	}
}

func (i *Instance) writeTempKeys() (string, string, error) {
	keyFile, err := os.CreateTemp("", "pullpreview-key-*")
	if err != nil {
		return "", "", err
	}
	if _, err := keyFile.WriteString(i.Access.PrivateKey + "\n"); err != nil {
		_ = keyFile.Close()
		return "", "", err
	}
	if err := keyFile.Close(); err != nil {
		return "", "", err
	}
	if err := os.Chmod(keyFile.Name(), 0600); err != nil {
		return "", "", err
	}

	certFile := ""
	if strings.TrimSpace(i.Access.CertKey) != "" {
		certFile = keyFile.Name() + "-cert.pub"
		if err := os.WriteFile(certFile, []byte(i.Access.CertKey+"\n"), 0600); err != nil {
			return "", "", err
		}
	}

	return keyFile.Name(), certFile, nil
}

func (i *Instance) EnsureRemoteAuthorizedKeysOwner() error {
	homeDir := HomeDirForUser(i.Username())
	command := fmt.Sprintf("chown %s:%s %s/.ssh/authorized_keys && chmod 0600 %s/.ssh/authorized_keys", i.Username(), i.Username(), homeDir, homeDir)
	return i.SSH(command, nil)
}

func (i *Instance) DeployApp(appPath string) error {
	switch i.DeploymentTarget {
	case DeploymentTargetHelm:
		return i.DeployWithHelm(appPath)
	case DeploymentTargetCompose:
		return i.DeployWithDockerContext(appPath)
	default:
		return fmt.Errorf("unsupported deployment target %q", i.DeploymentTarget)
	}
}

func (i *Instance) SetupScripts() error {
	if err := i.SetupSSHAccess(); err != nil {
		return err
	}
	if len(i.SSHPublicKeys()) > 0 {
		if err := i.EnsureRemoteAuthorizedKeysOwner(); err != nil {
			return err
		}
	}
	return nil
}

func (i *Instance) CloneIfURL(appPath string) (string, func(), error) {
	if strings.HasPrefix(appPath, "http://") || strings.HasPrefix(appPath, "https://") {
		parts := strings.SplitN(appPath, "#", 2)
		gitURL := parts[0]
		ref := "master"
		if len(parts) == 2 && parts[1] != "" {
			ref = parts[1]
		}
		tmpDir, err := os.MkdirTemp("", "pullpreview-app-*")
		if err != nil {
			return "", nil, err
		}
		cleanup := func() { _ = os.RemoveAll(tmpDir) }
		cmd := exec.CommandContext(i.Context, "git", "clone", gitURL, "--depth=1", "--branch", ref, tmpDir)
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr
		if err := i.Runner.Run(cmd); err != nil {
			cleanup()
			return "", nil, err
		}
		return tmpDir, cleanup, nil
	}
	return appPath, func() {}, nil
}

func (i *Instance) AppSizeMB(path string) float64 {
	info, err := os.Stat(path)
	if err != nil {
		return 0
	}
	return float64(info.Size()) / 1024.0 / 1024.0
}
